---
title: "Volcanoes"
author: "Imogen"
date: "2023-11-17"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(pheatmap)
library(limma)
library(DESeq2)
library(clusterProfiler)
library(mgsa)
library(DBI)
library(enrichplot)
library(UpSetR)
library(apeglm)
library(tibble)
library(EnhancedVolcano)
library(ggbiplot)

```

## Grabbing data from Fares's code

```{r}
### Read in the raw gene counts
t10echlorRep1_raw <- read_tsv("10d1ReadsPerGene.out.tab", skip = 4, col_names = F)
t10echlorRep2_raw <- read_tsv("10d2ReadsPerGene.out.tab", skip = 4, col_names = F)
t10echlorRep3_raw <- read_tsv("10d3ReadsPerGene.out.tab", skip = 4, col_names = F)
apoechlorRep1_raw <- read_tsv("apo1ReadsPerGene.out.tab", skip = 4, col_names = F)
apoechlorRep2_raw <- read_tsv("apo2ReadsPerGene.out.tab", skip = 4, col_names = F)
apoechlorRep3_raw <- read_tsv("apo3ReadsPerGene.out.tab", skip = 4, col_names = F)

### Rename columns (4 columns)
colnames(t10echlorRep1_raw) <- c("geneID","totalReads","+","-")
colnames(t10echlorRep2_raw) <- c("geneID","totalReads","+","-")
colnames(t10echlorRep3_raw) <- c("geneID","totalReads","+","-")
colnames(apoechlorRep1_raw) <- c("geneID","totalReads","+","-")
colnames(apoechlorRep2_raw) <- c("geneID","totalReads","+","-")
colnames(apoechlorRep3_raw) <- c("geneID","totalReads","+","-")

### Convert table to long format to filter out the 'totalReads' 
### column (kind of counterproductive since I sum them later anyways)

t10echlorRep1Tidy_raw <- t10echlorRep1_raw %>%
  dplyr::select(-totalReads) %>%
  pivot_longer(cols = c(2,3), names_to = "strand", values_to = "t10R1")
t10echlorRep2Tidy_raw <- t10echlorRep2_raw %>%
  dplyr::select(-totalReads) %>%
  pivot_longer(cols = c(2,3), names_to = "strand", values_to = "t10R2")
t10echlorRep3Tidy_raw <- t10echlorRep3_raw %>%
  dplyr::select(-totalReads) %>%
  pivot_longer(cols = c(2,3), names_to = "strand", values_to = "t10R3")
apoechlorRep1Tidy_raw <- apoechlorRep1_raw %>%
  dplyr::select(-totalReads) %>%
  pivot_longer(cols = c(2,3), names_to = "strand", values_to = "apoR1")
apoechlorRep2Tidy_raw <- apoechlorRep2_raw %>%
  dplyr::select(-totalReads) %>%
  pivot_longer(cols = c(2,3), names_to = "strand", values_to = "apoR2")
apoechlorRep3Tidy_raw <- apoechlorRep3_raw %>%
  dplyr::select(-totalReads) %>%
  pivot_longer(cols = c(2,3), names_to = "strand", values_to = "apoR3")

### Merge all 6 files into a merged table/dataset
mergedEchlorTidy_raw <- merge(t10echlorRep1Tidy_raw,t10echlorRep2Tidy_raw) %>%
  merge(t10echlorRep3Tidy_raw) %>%
  merge(apoechlorRep1Tidy_raw) %>%
  merge(apoechlorRep2Tidy_raw) %>%
  merge(apoechlorRep3Tidy_raw)

echlorGTF_data <- read_tsv("GCA_003991915.1_ElyChl2.0_genomic.gtf", skip = 3, col_names = F)
colnames(echlorGTF_data) <- c("accession", "source", "feature","start","end","spacer","strand","spacer2","description")
echlorGeneGTF_data <- echlorGTF_data %>% filter(feature == "gene")
echlorGeneGTF_data$geneID <- str_match(echlorGeneGTF_data$description, '(EGW[_0-9]*)')[,1]
featAnnotations <- echlorGeneGTF_data[,c("accession","start","end","strand","geneID")]

temp <- mergedEchlorTidy_raw %>% merge(featAnnotations, by = c("geneID", "strand"), all.x = TRUE) %>%
  mutate(length = end - start)

temp1 <- temp %>% filter(is.na(accession))
temp2 <- temp %>% filter(!is.na(accession))
temp2['t10R1'] <- temp2['t10R1'] + temp1['t10R1']
temp2['t10R2'] <- temp2['t10R2'] + temp1['t10R2']
temp2['t10R3'] <- temp2['t10R3'] + temp1['t10R3']
temp2['apoR1'] <- temp2['apoR1'] + temp1['apoR1']
temp2['apoR2'] <- temp2['apoR2'] + temp1['apoR2']
temp2['apoR3'] <- temp2['apoR3'] + temp1['apoR3']

## Normalize by read length (calculated earlier by end_position - start_position)
mergedEchlorTidy_RPK <- mergedEchlorTidy_processed %>%
  mutate(across(t10R1:apoR3, function(x) x/length))

## Calculate scaling factor (this is the sum of length-normalized reads in each sample. 
## Divide by a million for TPM)
echlor_scaling <- mergedEchlorTidy_RPK[,c(3:8)] %>%
  apply(2, function(x) sum(x, na.rm = T)/10^6)

## Divide all values by the corresponding scaling factor. Sweep function applies a calculation
## between table A and table B where tableB is just one row.
mergedEchlorNorm_TPM <- mergedEchlorTidy_RPK
mergedEchlorNorm_TPM[,c(3:8)] <- sweep(mergedEchlorNorm_TPM[,c(3:8)], 2, echlor_scaling, "/")

mergedEchlorTidy_processed <- temp2

mergedEchlorTidy_RPK <- mergedEchlorTidy_processed %>%
  mutate(across(t10R1:apoR3, function(x) x/length))
mergedEchlorNorm_TPM <- mergedEchlorTidy_RPK
mergedEchlorNorm_TPM[,c(3:8)] <- sweep(mergedEchlorNorm_TPM[,c(3:8)], 2, echlor_scaling, "/")

## Sanity check to see most abundantly expressed t10R1 gene
arrange(mergedEchlorNorm_TPM, desc(t10R1))

```

## Generating DESeq object
``` {r}
sampleTable <- data.frame(Sample = c("Sample1", "Sample2", "Sample3"),
  Condition = c("d10", "d10", "d10", "apo","apo","apo"))

#Read gene function data
gene_function_file <- "uniprotkb_elysia_chlorotica_2023_10_30.tsv"
gene_function_info <- read.delim(gene_function_file,header=TRUE,sep="\t") 
gene_function_info <- gene_function_info[order(gene_function_info$Gene.Names),]

# Create a count matrix with genes as rows and samples as columns
countMatrix <- mergedEchlorTidy_processed %>%
  column_to_rownames('geneID') %>%
  dplyr::select(2,3,4,5,6,7) %>%
  as.matrix()
colnames(countMatrix) <- c(rep("d10",3),rep("apo",3))

### Create a DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = round(countMatrix), colData = sampleTable, design = ~ Condition) 
dds <- DESeq(dds)
res <- results(dds, name="Condition_d10_vs_apo")
resLFC <- lfcShrink(dds, coef="Condition_d10_vs_apo", type="apeglm")
resOrdered <- na.omit(res[order(res$padj),])
```


## Generating PCA plot using prcomp
```{r, echo=FALSE}
countMatrix <- mergedEchlorNorm_TPM %>%
  column_to_rownames('geneID') %>%
  dplyr::select(2,3,4,5,6,7) %>% as.matrix()
colnames(countMatrix) <- c(rep("d10",3),rep("apo",3))
dds.pcoa <- countMatrix %>% t() %>% prcomp()
g <- ggbiplot(pcobj = dds.pcoa, var.axes = FALSE,groups=c(rep("d10",3),rep("apo",3)))+
  ggtitle("PCA from prcomp") + coord_equal(ratio = 0.5)
g
```

## Pulling KEGG data
```{r, echo=FALSE}
echlorGene_data <- read_delim("GCA_003991915.1_ElyChl2.0_rna_from_genomic.fna",col_names = F,delim = "ยง") %>%
  mutate(geneID = str_match(X1, "locus_tag=([0-9A-Za-z_]*)")[,2],
         accession = str_match(X1, "lcl\\|([0-9A-Za-z\\._]*)")[,1]) %>%
  dplyr::select(geneID, accession) %>% na.omit()

echlor_Kegg_data <- read_tsv("eChlor_KEGGOut.txt",col_names = FALSE)
echlor_Kegg_data <- plyr::rename(echlor_Kegg_data, c("X1" = "accession", "X2" = "ko_term")) %>%
  merge(echlorGene_data,all.x = TRUE) %>%
  dplyr::select(2,3)

echlor_Kegg_data <- echlor_Kegg_data %>%
  filter(!is.na(ko_term))

## TERM2GENE (tibble that maps K term to pathway - needed by enricher)
keggTemp <- bitr_kegg(echlor_Kegg_data$ko_term,"kegg","Path","ko") %>%
  dplyr::select(2,1)

temp <- echlor_Kegg_data %>% dplyr::filter(geneID %in% d10vsApo_hits$Row.names)

keggResults <- enrichKEGG(gene=temp$ko_term,
           organism ="ko",
           pvalueCutoff = 0.05,
           universe=echlor_Kegg_data$ko_term,
           pAdjustMethod = "BH")

pathwayOfInterest <- "ko00010" 
tempKegg_res <- as.data.frame(keggResults)
keggPathToTerm <- bitr_kegg(echlor_Kegg_data$ko_term,"kegg","Path","ko") %>%
  dplyr::select(2,1)

## Filter K-terms that map to our pathway of interest
ktermsInPath <- dplyr::filter(keggPathToTerm,Path==pathwayOfInterest) %>%
  dplyr::pull(kegg)
## There is some redundancy - i.e. there are multiple E. chlor genes that map to the same k-term
genesInPath <- dplyr::filter(echlor_Kegg_data,ko_term%in%ktermsInPath) %>% pull(geneID)
```

## Generating volcano plot
```{r, echo=FALSE}
volcano_data <- data.frame(log2FoldChange=res$log2FoldChange, log10P=-log10(res$padj), gene = rownames(res))
volcano_data$inPathway <- ifelse(volcano_data$gene %in% genesInPath, TRUE,FALSE) 
volcano_plot <- ggplot(volcano_data%>%arrange(inPathway), aes(x = log2FoldChange, y = log10P, color=inPathway)) +
    geom_point(alpha=0.5) + theme_minimal() +
    scale_color_manual(values = c("FALSE" = "grey", "TRUE"="red")) +  
    labs(title = "Volcano Plot",x = "log2(Fold Change)",y = "-log10(Adjusted P-value)",color="In Pathway")
volcano_plot
```

